core/src/main/scala/cats/instances/list.scala

```diff
@@ -156,8 +156,8 @@ trait ListInstances extends cats.kernel.instances.ListInstances {
             // traversing fa, which we will do fully
             // in all cases.
             Eval.always {
-              val gb = f(a)
-              G.void(gb)
+              val fb = f(a)
+              G.void(fb)
```

Comment:
we don't need this now. `fb: G[Unit]` so the void is a no-op.

Reply:
We absolutely do, or the change is not compatible. See https://github.com/typelevel/cats/pull/4352#discussion_r1027255402.

Reply:
I don't follow you. My claim is that `f(a): G[Unit]` already. While `G.void(f(a))` is still `G[Unit]`, it isn't needed and is likely to be wasteful.

In fact, unconditionally calling `void` internal to a method is a code smell that should have alerted us the type was poor to begin with.

That said, there could be ergonomic reasons to not have the users required to call this void.

Reply:
> I don't follow you. My claim is that `f(a): G[Unit]` already

Except, it's not really :) sure, that's what the current signature claims, and a signature of `G[Unit]` is binary-compatible with a signature of `G[A]` for an unbounded type parameter `A`.

But Cats still has to retain compatibility with calling code that was passing e.g. a `G[String]` there. Otherwise we will get class cast exceptions, exactly as demonstrated in https://github.com/typelevel/cats/pull/4352#discussion_r1027255402.

Reply:
@johnynek just to spell it out: once the MIMA test was added, returning `f(a)` without void caused it to break.

Reply:
Sorry to be pedantic :) actually, that's not a MiMa test. MiMa compares the binary-compatibility of signatures. MiMa can only detect linking errors but not other errors such as class casting exceptions. In this case there is no linking error, because due to type erasure the method signature is the same.

What I added was a runtime-compatibility test. Cats has an internal test project, that is _compiled_ against Cats 2.0.0 but _run_ against the latest development version of Cats. It shows that code compiled with an old version of Cats, that is not passing `G[Unit]` to these methods will encounter a class cast exception if we rewrite the code here to assume we are indeed getting a `G[Unit]`.

Reply:
In fact, I can make the exact point here. The whole point of `traverse_` is the avoid the boilerplate of calling `void`, so if I have to do `list.traverse_(a => f(a).void)`, why don't I just do `list.traverse(f).void`? (And if the answer is, to avoid allocating the list, if you really care about that there is still `foldMapA`)